File: src/main/java/com/kiwi/mobile/retrograph/ResultObservable.java
package com.kiwi.mobile.retrograph;

import io.reactivex.*;
import io.reactivex.disposables.*;
import io.reactivex.exceptions.*;
import io.reactivex.plugins.*;

import retrofit2.*;

final class ResultObservable<T> extends Observable<Result<T>> {
  private final Observable<Response<T>> upstream;

  ResultObservable(Observable<Response<T>> upstream) {
    this.upstream = upstream;
  }

  @Override
  protected void subscribeActual(Observer<? super Result<T>> observer) {
    upstream.subscribe(new ResultObserver<T>(observer));
  }

  private static class ResultObserver<R> implements Observer<Response<R>> {
    private final Observer<? super Result<R>> observer;

    ResultObserver(Observer<? super Result<R>> observer) {
      this.observer = observer;
    }

    @Override
    public void onSubscribe(Disposable disposable) {
      observer.onSubscribe(disposable);
    }

    @Override
    public void onNext(Response<R> response) {
      observer.onNext(Result.response(response));
    }

    @Override
    public void onError(Throwable throwable) {
      try {
        observer.onNext(Result.<R>error(throwable));
      } catch (Throwable t) {
        try {
          observer.onError(t);
        } catch (Throwable inner) {
          Exceptions.throwIfFatal(inner);
          RxJavaPlugins.onError(new CompositeException(t, inner));
        }
        return;
      }
      observer.onComplete();
    }

    @Override
    public void onComplete() {
      observer.onComplete();
    }
  }
}

File: src/main/java/com/kiwi/mobile/retrograph/rxjava/ResultObservable.kt
package com.kiwi.mobile.retrograph.rxjava

import io.reactivex.*
import io.reactivex.disposables.*
import io.reactivex.exceptions.*
import io.reactivex.plugins.*

import retrofit2.*
import retrofit2.adapter.rxjava2.*

class ResultObservable<T>(
  private val upstream: Observable<Response<T>>
):
  Observable<Result<T>>() {

  // region Private Types

  private class ResultObserver<R>(
    private val observer: Observer<in Result<R>>
  ):
    Observer<Response<R>> {

    // region Private Properties

    private var terminated: Boolean = false

    // endregion Private Properties

    // region Public Methods

    override fun onSubscribe(disposable: Disposable) {
      sendSubscribe(disposable)
    }

    override fun onNext(response: Response<R>) {
      sendResponse(response)
    }

    override fun onError(throwable: Throwable) {
      sendError(throwable)
      sendComplete()
    }

    override fun onComplete() {
      sendComplete()
    }

    // endregion Public Methods

    // region Private Methods

    private fun sendSubscribe(disposable: Disposable) {
      observer.onSubscribe(disposable)
    }

    private fun sendResponse(response: Response<R>) {
      observer.onNext(Result.response(response))
    }

    private fun sendError(throwable: Throwable) {
      try {
        observer.onNext(Result.error(throwable))
      } catch (nextThrowable: Throwable) {
        throwException(nextThrowable)
        return
      }
    }

    private fun sendComplete() {
      if (!terminated) {
        terminated = true
        observer.onComplete()
      }
    }

    private fun throwException(throwable: Throwable) {
      try {
        terminated = true
        observer.onError(throwable)
      } catch (innerThrowable: Throwable) {
        Exceptions.throwIfFatal(innerThrowable)
        RxJavaPlugins.onError(CompositeException(throwable, innerThrowable))
      }
    }

    // endregion Private Methods
  }

  // endregion Private Types

  // region Protected Methods

  override fun subscribeActual(observer: Observer<in Result<T>>) {
    upstream.subscribe(ResultObserver(observer))
  }

  // endregion Protected Methods
}
